pkg/operator/workloads/metrics.go

```diff
@@ -0,0 +1,463 @@
+/*
+Copyright 2019 Cortex Labs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workloads
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+	"github.com/cortexlabs/cortex/pkg/lib/parallel"
+	"github.com/cortexlabs/cortex/pkg/lib/slices"
+	"github.com/cortexlabs/cortex/pkg/operator/api/context"
+	"github.com/cortexlabs/cortex/pkg/operator/api/schema"
+	"github.com/cortexlabs/cortex/pkg/operator/api/userconfig"
+	"github.com/cortexlabs/cortex/pkg/operator/config"
+)
+
+const (
+	maxClasses = 75 // GeMetricData can get up to 100 metrics per request, avoid multiple requests and have room for other stats
+)
+
+func GetMetrics(appName, apiName string) (*schema.APIMetrics, error) {
+	ctx := CurrentContext(appName)
+	api := ctx.APIs[apiName]
+
+	apiSavedStatus, err := getAPISavedStatus(api.ID, api.WorkloadID, appName)
+	if err != nil {
+		return nil, err
+	}
+
+	if apiSavedStatus == nil {
+		return nil, errors.Wrap(ErrorNotFound(), apiName)
+	}
+
+	if apiSavedStatus.Start == nil {
+		return nil, errors.Wrap(ErrorNotFound(), apiName, "start time")
+	}
+
+	apiStartTime := apiSavedStatus.Start.Truncate(time.Second)
+
+	// Get realtime metrics for the seconds elapsed in the latest minute
+	realTimeEnd := time.Now().Truncate(time.Second)
+	realTimeStart := realTimeEnd.Truncate(time.Minute)
+
+	realTimeMetrics := schema.APIMetrics{}
+	batchMetrics := schema.APIMetrics{}
+
+	requestList := []func() error{}
+	if realTimeStart.Before(realTimeEnd) {
+		requestList = append(requestList, getAPIMetrics(appName, api, 1, &realTimeStart, &realTimeEnd, &realTimeMetrics))
+	}
+
+	if apiStartTime.Before(realTimeStart) {
+		batchEnd := realTimeStart
+		twoWeeksAgo := batchEnd.Add(-14 * 24 * time.Hour)
+
+		var batchStart time.Time
+		if twoWeeksAgo.Before(*apiSavedStatus.Start) {
+			batchStart = *apiSavedStatus.Start
+		} else {
+			batchStart = twoWeeksAgo
+		}
+		requestList = append(requestList, getAPIMetrics(appName, api, 60*60, &batchStart, &batchEnd, &batchMetrics))
+	}
+
+	if len(requestList) != 0 {
```

Comment:
if `len(requestList) != 0` should this return `nil` or an error?

Reply:
I am not sure what you mean by this